[Remote Clusters] Add IPv6 support

kapral18 opened the pull request:
fixes #228298

- Add IPv6 address support for remote cluster seed nodes and proxy addresses using bracket notation ( [::1]:9300 )
- Move address validation from client-side to shared common library using  ipaddr.js  for robust IPv4/IPv6 validation
- Update cluster serialization and cloud URL parsing to handle IPv6 addresses correctly

## How to test

1. Spin up local kibana and es on default ports (`yarn es snapshot` and `yarn start --no-base-path`)
2. Go to `/app/management/data/remote_clusters`
3. Click "add new cluster"
4. Click "certificates celect"
5. Add loopback ipv6 address to seed pointing to cluster communication port `[::1]:9300`
6. Observe no client side error for `seed` field, proving that ipv6 syntax is now valid
7. Add `test` name
8. `Uncheck skip_unavailable`
10. Click `next`
11. Click `create remote cluster`
12. Observe live cluster connection in the list of remote clusters

## Release Note
Added IPv6 support to address fields in Remote Clusters UI

<!--ONMERGE {"backportTargets":["8.19","9.1"]} ONMERGE-->

kapral18 committed:
chore: initial draft

kapral18 committed:
chore: adopt existing ipaddr.js

- Comment out failing test case for IPv4-mapped IPv6 due to unsupported format
- Replace custom IPv4, IPv6, and hostname validation logic with `ipaddr.js`
- Add helper function `looksLikeIPv4` to identify numeric IPv4 attempts
- Update `parseAddressAndPort` to handle IPv6 without brackets and improve port parsing
- Simplify `isAddressValid` to leverage `ipaddr.js` for IPv4 and IPv6 validation
- Ensure `isPortValid` correctly checks for both address and port presence

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

kapral18 committed:
chore: update validation to support IPv4, IPv6, and hostname formats

- Updated tests and validation logic to support IPv4, IPv6, and hostname formats for seed nodes and proxy addresses.
- Added new snapshot tests for IPv4 and IPv6 address validation.
- Modified `extractHostAndPort` function to handle IPv6 addresses with and without brackets.
- Updated error messages to reflect the new supported formats.
- Enhanced `validateProxy` and `validateSeed` functions to correctly parse and validate IPv6 addresses.
- Adjusted `convertCloudRemoteAddressToProxyConnection` to handle different URL formats and default ports.

kapral18 committed:
chore: update cluster serialization to handle different proxy address formats

- Refactor test cases to include hostname, IPv4, and IPv6 proxy addresses.
- Modify `deserializeCluster` to use `extractHostAndPort` for extracting the host from the deprecated proxy address.
- Update `isCloudAdvancedOptionsEnabled` to correctly compare the extracted host with the server name for all address types.
- Move `validate_address` and related functions to `common/lib` for better reusability.
- Adjust imports in various files to use the new location of `validate_address` functions.

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/334120)
* Commit: 3b7bb19b5ee88ca624f07a8c04507a4403adc273

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/3adac4b3a2db19a32f01356248fca04abe589030) | [after](https://github.com/elastic/kibana/commit/3b7bb19b5ee88ca624f07a8c04507a4403adc273) | diff |
| --- | --- | --- | --- |
| `remoteClusters` | 98 | 99 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/3adac4b3a2db19a32f01356248fca04abe589030) | [after](https://github.com/elastic/kibana/commit/3b7bb19b5ee88ca624f07a8c04507a4403adc273) | diff |
| --- | --- | --- | --- |
| `remoteClusters` | 71.4KB | 82.5KB | +11.1KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/3adac4b3a2db19a32f01356248fca04abe589030) | [after](https://github.com/elastic/kibana/commit/3b7bb19b5ee88ca624f07a8c04507a4403adc273) | diff |
| --- | --- | --- | --- |
| `remoteClusters` | 9.0KB | 9.0KB | +15.0B |



cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/334120","number":334120,"commit":"3b7bb19b5ee88ca624f07a8c04507a4403adc273"}],"number":334120}
buildkite-pr-comment-->

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into enhancement/remote-clusters/support-ipv6-in-remote-address-field

sabarasaba reviewed (approved):
thanks for working on this @kapral18, code changes lgtm tested locally 🚀

kapral18 commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/233682">](https://github.com/elastic/kibana/pull/233682)|
|❌|8.19|An unhandled error occurred. Please see the logs for details|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 233415
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.1 -->

kapral18 commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/233683">](https://github.com/elastic/kibana/pull/233683)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.1 -->
